src/NServiceBus.Gateway/Receiving/SingleCallChannelReceiver.cs

```diff
@@ -93,13 +108,32 @@ async Task HandleSubmit(CallInfo callInfo)
                 await stream.ReadAsync(body, 0, body.Length).ConfigureAwait(false);
                 args.Body = body;
 
-                if (await deduplicator.DeduplicateMessage(callInfo.ClientId, DateTime.UtcNow, new ContextBag()).ConfigureAwait(false))
+                var context = new ContextBag();
+                if (await deduplicationStorage.IsDuplicate(callInfo.ClientId, context).ConfigureAwait(false))
                 {
-                    await messageReceivedHandler(args).ConfigureAwait(false);
+                    Logger.InfoFormat("Message with id: {0} is already on the bus, dropping the request", callInfo.ClientId);
                 }
                 else
                 {
-                    Logger.InfoFormat("Message with id: {0} is already on the bus, dropping the request", callInfo.ClientId);
+                    await messageReceivedHandler(args).ConfigureAwait(false);
+                    try
+                    {
+                        await deduplicationStorage.MarkAsDispatched(callInfo.ClientId, context).ConfigureAwait(false);
+                    }
+                    catch (Exception e)
+                    {
+                        // swallow exception in non-dtc modes.
+                        // When using no transactions, the message has been sent to the transport already. Throwing would cause the operation to be retried and a guaranteed duplicate to be created. By swallowing the exception, the duplicate is only created if the same message is sent to the gateway for another reason.
+                        // When using distributed transactions, throw so that both persistence and transport can rollback atomically.
+                        if (useTransactionScope)
+                        {
+                            throw;
+                        }
+                        else
+                        {
+                            Logger.Warn($"Failed to mark message with id '{callInfo.ClientId}' as dispatched. This message won't be deduplicated.'", e);
```

Comment:
Why is this called ClientId and not just MessageIdentifier or MessageID?

Reply:
I have no idea where this naming is coming from. From my understanding it's just the original message id from the sending site but no idea why this is named ClientId. @andreasohlund @janovesk any ideas?

Reply:
It's just always been that way. My hunch is that its due to the fact that we allow webclients to send raw messages via the gateway

https://github.com/Particular/docs.particular.net/blob/master/samples/gateway/Gateway_3/WebClient/Index.html#L16

so that might be where this comes from

Reply:
client id would still be highly misleading in that case as a client would probably not change it's "client id" for every message it dispatches.

Reply:
agreed